NachoSpigot-Server/src/main/java/dev/cobblesword/nachospigot/hitdetection/LagCompensator.java

```diff
@@ -0,0 +1,96 @@
+package dev.cobblesword.nachospigot.hitdetection;
+
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.ListMultimap;
+import me.elier.nachospigot.config.NachoConfig;
+import org.apache.commons.lang3.tuple.Pair;
+import org.bukkit.Location;
+import org.bukkit.entity.Player;
+import org.bukkit.util.Vector;
+
+import java.util.*;
+
+public class LagCompensator {
+
+    /*
+        Lag Compensator based on https://github.com/Islandscout/ServerSideHitDetection
+        Adjusts hit detection to be more fair for players with worse ping.
+     */
+
+    private final ListMultimap<UUID, Pair<Location, Long>> locationTimes;
+    private final int historySize;
+    private final int pingOffset;
+    private final int timeResolution;
+
+    public LagCompensator(int historySize, int pingOffset, int timeResolution) {
+        this.locationTimes = ArrayListMultimap.create();
+        this.historySize = historySize;
+        this.pingOffset = pingOffset;
+        this.timeResolution = timeResolution;
+    }
+
+    public Location getHistoryLocation(int rewindMillisecs, Player player) {
+        if (!locationTimes.containsKey(player.getUniqueId()))
+            return player.getLocation();
+
+        List<Pair<Location, Long>> times = locationTimes.get(player.getUniqueId());
+        long currentTime = System.currentTimeMillis();
+
+        int rewindTime = rewindMillisecs + pingOffset;
+        int timesSize = times.size() - 1;
```

Comment:
Question, is `times.size()` accessed from multiple threads? If so, watch out for data races. They are a pain in the ass to bug-find. If so, look for something thread-safe. If not, keep it like this :)

Reply:
I'm not precisely sure, since I'm not fully aware of how threading is done in the server. This method is only called when the server receives a Entity Use packet through a method from the Entity class.

Reply:
Then this is fine for now. If async packets may be added in the future we will have to keep this in mind.